lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -70,45 +102,138 @@ defaultManagerSettings =
 newManager :: MonadIO m => ManagerSettings -> m Manager
 newManager = HTTPS.newTlsManagerWith
 
+-- | Simply return a pool metadata url, unchanged
+identityUrlBuilder
+    :: StakePoolMetadataUrl
+    -> StakePoolMetadataHash
+    -> Either HttpException URI
+identityUrlBuilder (StakePoolMetadataUrl url) _ =
+    maybe (Left e) Right $ parseURI (T.unpack url)
+  where
+    e = InvalidUrlException (T.unpack url) "Invalid URL"
+
+-- | Build a URL from a metadata hash compatible with an aggregation registry
+registryUrlBuilder
+    :: URI
+    -> StakePoolMetadataUrl
+    -> StakePoolMetadataHash
+    -> Either HttpException URI
+registryUrlBuilder baseUrl _ (StakePoolMetadataHash bytes) =
+    Right $ baseUrl
+        { uriPath = "/" <> intercalate "/" (pathSegments baseUrl ++ [hash])
+        }
+  where
+    hash = T.unpack $ T.decodeUtf8 $ convertToBase Base16 bytes
+
 fetchFromRemote
-    :: Manager
+    :: Tracer IO StakePoolMetadataFetchLog
+    -> [StakePoolMetadataUrl -> StakePoolMetadataHash -> Either HttpException URI]
+    -> Manager
     -> StakePoolMetadataUrl
     -> StakePoolMetadataHash
-    -> IO (Either String StakePoolMetadata)
-fetchFromRemote manager url_ hash_ = runExceptT $ do
-    req <- fromInvalidUrl $ parseUrlThrow (T.unpack url)
-    chunk <- ExceptT
-        $ handle fromIOException
-        $ handle fromHttpException
-        $ withResponse req manager $ \res -> do
-        -- NOTE
-        -- Metadata are _supposed to_ be made of:
-        --
-        -- - A name (at most 50 UTF-8 bytes)
-        -- - An optional description (at most 250 UTF-8 bytes)
-        -- - A ticker (at most 5 UTF-8 bytes)
-        -- - A homepage (at most 100 UTF-8 bytes)
-        --
-        -- So, the total, including a pretty JSON encoding with newlines ought
-        -- to be less than 512 bytes. For security reasons, we only download the
-        -- first 512 bytes.
-        pure . BL.toStrict <$> brReadSome (responseBody res) 512
-    when (blake2b256 chunk /= hash) $ throwE $ mconcat
+    -> IO (Maybe StakePoolMetadata)
+fetchFromRemote tr builders manager url hash = runExceptTLog $ do
+    chunk <- getChunk `fromFirst` builders
+    when (blake2b256 chunk /= coerce hash) $ throwE $ mconcat
         [ "Metadata hash mismatch. Saw: "
         , B8.unpack $ hex $ blake2b256 chunk
         , ", but expected: "
-        , B8.unpack $ hex hash
+        , B8.unpack $ hex $ coerce @_ @ByteString hash
         ]
     except $ eitherDecodeStrict chunk
   where
-    StakePoolMetadataUrl  url  = url_
-    StakePoolMetadataHash hash = hash_
+    runExceptTLog
+        :: ExceptT String IO StakePoolMetadata
+        -> IO (Maybe StakePoolMetadata)
+    runExceptTLog action = runExceptT action >>= \case
+        Left msg ->
+            Nothing <$ traceWith tr (MsgFetchPoolMetadataFailure hash msg)
+
+        Right meta ->
+            Just meta <$ traceWith tr (MsgFetchPoolMetadataSuccess hash meta)
```

Comment:
Previously I have combined Success/Failure log messages into a single message type, e.g.
```
    | MsgFetchPoolMetadataResult StakePoolMetadataHash (Either String StakePoolMetadata)
```
This way is no less expressive for logging, and means you can use a single `traceWith` statement and avoid needing to use constructs like `runExceptTLog`.